core/lib/src/response/response.rs

```diff
@@ -522,46 +573,57 @@ impl<'r> ResponseBuilder<'r> {
     /// let custom_values: Vec<_> = response.headers().get("X-Custom").collect();
     /// assert_eq!(custom_values, vec!["value 2", "value 3", "value 1"]);
     /// # }
+    /// 
+    /// # })
     /// ```
     #[inline(always)]
     pub fn join(&mut self, other: Response<'r>) -> &mut ResponseBuilder<'r> {
         self.response.join(other);
         self
     }
 
-    /// Retrieve the built `Response`.
-    ///
-    /// # Example
-    ///
-    /// ```rust
-    /// use rocket::Response;
-    ///
-    /// # #[allow(unused_variables)]
-    /// let response = Response::build()
-    ///     // build the response
-    ///     .finalize();
-    /// ```
-    #[inline(always)]
-    pub fn finalize(&mut self) -> Response<'r> {
-        std::mem::replace(&mut self.response, Response::new())
-    }
-
     /// Retrieve the built `Response` wrapped in `Ok`.
     ///
     /// # Example
     ///
     /// ```rust
     /// use rocket::Response;
+    /// 
+    /// # rocket::async_test(async {
     ///
     /// let response: Result<Response, ()> = Response::build()
     ///     // build the response
-    ///     .ok();
+    ///     .ok().await;
     ///
     /// assert!(response.is_ok());
+    /// 
+    /// # })
     /// ```
     #[inline(always)]
-    pub fn ok<T>(&mut self) -> Result<Response<'r>, T> {
-        Ok(self.finalize())
+    pub async fn ok<E>(&mut self) -> Result<Response<'r>, E> {
+        Ok(self.await)
+    }
+}
+
+impl<'r> Future for ResponseBuilder<'r> {
+    type Output = Response<'r>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```

Comment:
I think we should keep `finalize()` but convert it to an `async fn`, primarily for future extensibility and readability.

Reply:
The idea behind making ResponseBuilder a future itself came from <https://blog.yoshuawuyts.com/async-finalizers/>, based on the idea that `.finalize().await` is kind of redundant. If you really want to keep it a separate function, I'll do it, but I'm not sure how it aids in extensibility (you can already see that other modes like `ok()` still work).

Reply:
I would be happy to tackle this part last. My main hesitation is that adding functionality to the manual implementation will be harder than adding it to an `async fn`, and implementing `Future` directly on `ResponseBuilder` is a strong commitment in comparison.

Reply:
My new implementation of it is actually based on an async block anyway, so it shouldn't be a problem.